### PR TITLE
Improving test coverage

### PR DESCRIPTION
**Pull request by dannym87**
- Adding unit tests for ArraySerializable
- Improving ClassMethodsTest to verify exceptions are thrown as expected
- Improving NumberOfParameterFilterTest to verify exceptions are thrown as expected
- Added ClassMethodsTraversableOptions so we can test setting ClassMethods options from a Traversable object

### CODE REVIEW

File: test/ClassMethodsTest.php [hunk 1]
```diff
@@ -73,4 +76,42 @@ public function testCanHydratedPromiscuousInstances()
...
+        $this->setExpectedException(InvalidArgumentException::class);
+        $this->hydrator->setOptions('invalid options');
+    }
+
+    /**
+     * Verifies options can be set from a Traversable object
+     */
+    public function testSetOptionsFromTraversable()
+    {
+        $options = new ClassMethodsTraversableOptions();
```
Comment (Maks3w) [inline test/ClassMethodsTest.php]: You could either use ArrayAccess or mock FilterIterator for mock the Iterator interface

**Comment by Maks3w** [comment]
Please add the expected exception message too

File: test/ClassMethodsTest.php (hunk 1 above)
Comment (dannym87) [inline test/ClassMethodsTest.php]: Removed ClassMethodsTraversableOptions test asset and used ArrayObject instead

File: test/ClassMethodsTest.php [hunk 2]
```diff
@@ -10,6 +10,9 @@
 namespace ZendTest\Hydrator;
 
 use Zend\Hydrator\ClassMethods;
+use Zend\Hydrator\Exception\BadMethodCallException;
+use Zend\Hydrator\Exception\InvalidArgumentException;
+use Zend\Stdlib\ArrayObject;
```
Comment (Maks3w) [inline test/ClassMethodsTest.php]: Use PHP native `ArrayObject`

Comment (dannym87) [inline test/ClassMethodsTest.php]: My mistake, thought I did use the native implementation. Fixed.

**Pull request closed without merging**